pkg/gce-pd-csi-driver/controller_test.go

```diff
@@ -1413,3 +1413,76 @@ func TestPickRandAndConsecutive(t *testing.T) {
 
 	}
 }
+
+func TestVolumeOperationConcurrency(t *testing.T) {
+	snapshotToExecute := make(chan gce.SnapshotSourceAndTarget, 3)
+	readyToExecute := make(chan struct{}, 2)
+	reqFinished := make(chan error, 2)
+
+	gceDriver := initBlockingGCEDriver(t, nil, snapshotToExecute, readyToExecute)
+	cs := gceDriver.cs
+
+	vol1CreateSnapshotReqA := &csi.CreateSnapshotRequest{
+		Name:           name + "1A",
+		SourceVolumeId: testVolumeId + "1",
+	}
+	vol1CreateSnapshotReqB := &csi.CreateSnapshotRequest{
+		Name:           name + "1B",
+		SourceVolumeId: testVolumeId + "1",
+	}
+	vol2CreateSnapshotReq := &csi.CreateSnapshotRequest{
+		Name:           name + "2",
+		SourceVolumeId: testVolumeId + "2",
+	}
+
+	runRequestInBackground := func(req *csi.CreateSnapshotRequest) {
+		_, err := cs.CreateSnapshot(context.Background(), req)
+		reqFinished <- err
+	}
+
+	sourceAndTargetForRequest := func(req *csi.CreateSnapshotRequest) gce.SnapshotSourceAndTarget {
+		volKey, err := common.VolumeIDToKey(req.SourceVolumeId)
+		if err != nil {
+			t.Fatalf("Could not find volume with ID %v: %v", req.SourceVolumeId, err)
+		}
+		return gce.SnapshotSourceAndTarget{
+			Source: volKey.String(),
+			Target: req.Name,
+		}
+	}
+
+	// Start first valid request vol1CreateSnapshotReqA and block until it reaches the CreateSnapshot
+	go runRequestInBackground(vol1CreateSnapshotReqA)
+	<-readyToExecute
+
+	// Check that vol1CreateSnapshotReqB is rejected, due to same volume ID
+	// Also allow vol1CreateSnapshotReqB to complete, in case it is allowed to Mount
+	snapshotToExecute <- sourceAndTargetForRequest(vol1CreateSnapshotReqB)
+	_, err := cs.CreateSnapshot(context.Background(), vol1CreateSnapshotReqB)
+	if err != nil {
+		serverError, ok := status.FromError(err)
+		if !ok {
+			t.Fatalf("Could not get error status code from err: %v", err)
+		}
+		if serverError.Code() != codes.Aborted {
+			t.Fatalf("Expected error code: %v, got: %v. err : %v", codes.Aborted, serverError.Code(), err)
+		}
+	} else {
+		t.Fatalf("Expected error: %v, got no error", codes.Aborted)
+	}
+
+	// Start second valid request vol2CreateSnapshotReq
+	go runRequestInBackground(vol2CreateSnapshotReq)
+
+	// Allow the vol2CreateSnapshotReq to complete, which it can concurrently with vol1CreateSnapshotReqA
+	snapshotToExecute <- sourceAndTargetForRequest(vol2CreateSnapshotReq)
+	if err = <-reqFinished; err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+
+	// To clean up, allow the vol1CreateSnapshotReqA to complete
+	snapshotToExecute <- sourceAndTargetForRequest(vol1CreateSnapshotReqA)
+	if err = <-reqFinished; err != nil {
```

Comment:
we should probably loop through `reqFinished` until completion at the end. In general I want to avoid code where the implementer/reviewer has to carefully count the number of sends on a channel and hand-write a corresponding number of receives on a channel. This is prone to error and very difficult to reason about